Replace 'Login with...' page with auto-redirect script

aaga opened the pull request:
There is also `localStorage` logic in here to support the "signup intent" feature (two separate buttons). That will be a separate PR on Elk though.

I used `localStorage` because I can't figure out how to prevent Mastodon from stripping all the URL params when redirecting to `/auth/sign_in`.

First part of https://mozilla-hub.atlassian.net/browse/SOCIALPLAT-104

aaga commented:
Screen cap of new flow:

https://github.com/MozillaSocial/mastodon/assets/7362472/5b742c6e-b622-48fe-9ef4-4e23a7cc8b12

toufali commented on app/views/auth/sessions/new.html.haml:
```diff
@@ -1,30 +1,22 @@
 
-- content_for :header_tags do
-  = render partial: 'shared/og'
+.simple_form.alternative-login
+  %h4= 'Redirecting...'
 
-- unless omniauth_only?
-  = simple_form_for(resource, as: resource_name, url: session_path(resource_name)) do |f|
-    %h1.title= t('auth.sign_in.title', domain: site_hostname)
-    %p.lead= t('auth.sign_in.preamble_html', domain: site_hostname)
-    .fields-group
-      - if use_seamless_external_login?
-        = f.input :email, autofocus: true, wrapper: :with_label, label: t('simple_form.labels.defaults.username_or_email'), input_html: { 'aria-label': t('simple_form.labels.defaults.username_or_email') }, hint: false
-      - else
-        = f.input :email, autofocus: true, wrapper: :with_label, label: t('simple_form.labels.defaults.email'), input_html: { 'aria-label': t('simple_form.labels.defaults.email') }, hint: false
-    .fields-group
-      = f.input :password, wrapper: :with_label, label: t('simple_form.labels.defaults.password'), input_html: { 'aria-label': t('simple_form.labels.defaults.password'), autocomplete: 'current-password' }, hint: false
-
-    .actions
-      = f.button :button, t('auth.login'), type: :submit
-
-- if devise_mapping.omniauthable? && resource_class.omniauth_providers.any?
```
Do we need to delete all of this?  Couldn't you just add your script under the `.simple_form.alternative-login` class?  I wonder if this might make managing upstream conflicts easier?

toufali reviewed (commented):
This works on my machine, and assuming no issues with other clients should suffice!

If you're interested in digging deeper, I imagine there should be a better way to hijack the flow in the controller, before we hit the view?  

I looked into it briefly and found that by creating a `new` method in `app/controllers/auth/sessions_controller.rb`, I was able to get into the flow before the view renders:

```
  def new
    redirect_to('https://mozilla.com')
    # HTTP.post('https://mastodon.mozsoc.local/auth/auth/openid_connect')
  end
```
Getting it to trigger the OIDC flow is of course another matter... we'd need to spend more time on it.

bassrock commented on app/views/auth/sessions/new.html.haml:
```diff
@@ -1,30 +1,22 @@
+  %h4= 'Redirecting...'
 
-- unless omniauth_only?
-  = simple_form_for(resource, as: resource_name, url: session_path(resource_name)) do |f|
```
Could we keep all this but add a new param instead?

aaga replied:
yes, I realized I can use the sso_redirect param that was defined in https://github.com/mastodon/mastodon/pull/26083, if I merge that one in

bassrock replied:
Yay!

bassrock replied:
Could actually merge this back upstream too with that.

aaga commented:
@toufali re: hijacking this flow earlier - I've gone down that rabbit hole a few times and unfortunately come up blank each time. My findings are:
- `/auth/auth/openid_connect` needs to be navigated to via POST, and I haven't found a way to do that without injecting a DOM script. Every other redirect capability only does GET.
- `/auth/auth/openid_connect` also has CSRF protection. This means (I think..) that it's expecting a CSRF authenticity token, with a corresponding mastodon cookie, to ensure that the navigation isn't via a maliciously forged link. And the only way I can see to set those is by a) "touching" a Mastodon page first for them to be set (which is basically what this solution does) or b) pulling that CSRF stuff into Elk somehow (which, 😬😬😬)

But I gently lay down the gauntlet to be picked up now or in the future!

aaga commented:
Addressed comments from above

@toufali I could not for the life of me figure out how to get `sso_redirect` from `InitialStateSerializer` into the haml file, so I redefined it in `ApplicationController`

I was aggressive about minimizing the HTML that is rendered in the redirect case, so that the page loads and redirects as fast as possible.
